### PR TITLE
agent/http: use `httptest.Server` for handler tests

### LINKED ISSUE #305: Homogenize the http proxy tests
https://github.com/grafana/xk6-disruptor/pull/274 introduces a new [test](https://github.com/grafana/xk6-disruptor/blob/main/pkg/agent/protocol/http/proxy_test.go#L455-L473) that uses  `httptest.Server` for setting the test. 

This approach defers from the approach used in other tests, increasing the complexity. 

If using `httpstest.Server` is suitable for other tests, use it to have a uniform way for setting the tests.

### PR DESCRIPTION
**Pull request by roobre**
# Description

Fixes #305 

This PR makes minor changes to http proxy handler tests so they are performed with `httptest.Server` instead of the fake client. This aligns existing tests with the newly added metric tests, and provides more coverage as implementation details of the whole Go HTTP stack is tested, instead of just the handler.

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works.   
- [x] I have run linter locally (`make lint`) and all checks pass.
- [x] I have run tests locally (`make test`) and all tests pass.
- [ ] I have run relevant e2e test locally (`make e2e-xxx` for `agent`, `disruptors`, `kubernetes` or `cluster` related changes)
- [ ] Any dependent changes have been merged and published in downstream modules<br>

### CODE REVIEW

File: pkg/agent/protocol/http/proxy_test.go [hunk 1]
```diff
@@ -371,35 +313,54 @@ func Test_ProxyHandler(t *testing.T) {
...
 			}
 
-			reqURL := fmt.Sprintf("http://%s%s", tc.config.ListenAddress, tc.path)
-			req := httptest.NewRequest(tc.method, reqURL, strings.NewReader(string(tc.body)))
-			recorder := httptest.NewRecorder()
-			handler.ServeHTTP(recorder, req)
-			resp := recorder.Result()
+			proxyServer := httptest.NewServer(handler)
+
+			req, err := http.NewRequest(tc.method, proxyServer.URL+tc.path, bytes.NewReader(tc.body))
```
Comment (roobre) [inline pkg/agent/protocol/http/proxy_test.go:344]: Reading this in the diff, it is not fully clear to me whether `tc.body` was intended to be the request body in the first place, specially given that all requests are (currently) `GET` requests. Should I get rid of this and add a new `requestBody` field to the test table?

Comment (pablochacin) [reply pkg/agent/protocol/http/proxy_test.go:344]: The `tc.body` field was intended to be used in requests that were not `GET`, which as you point out, are still missing in the test cases, but we should consider adding. 
 
What I don't understand is your proposal of getting rid of `tc.body` and adding a new `requestBody` field to the test table. Wouldn't it be just a rename of `body` by `requestBody`?

File: pkg/agent/protocol/http/proxy_test.go [hunk 2]
```diff
@@ -371,35 +313,54 @@ func Test_ProxyHandler(t *testing.T) {
...
 			if err != nil {
-				t.Errorf("error parsing upstream address %v", err)
-				return
+				t.Fatalf("error parsing httptest url")
 			}
+
 			handler := &httpHandler{
 				upstreamURL: *upstreamURL,
-				client:      client,
+				client:      http.DefaultClient,
```
Comment (pablochacin) [inline pkg/agent/protocol/http/proxy_test.go:337]: `client` was introduced only for testing. If we are now always using the DefaultClient, maybe we could simply remove this field.

File: pkg/agent/protocol/http/proxy_test.go (hunk 1 above)
Comment (roobre) [reply pkg/agent/protocol/http/proxy_test.go:344]: Maybe we're talking about different things, in the diff I see:

> `req := httptest.NewRequest(tc.method, reqURL, strings.NewReader(string(tc.body)))`

Which is the test case body being used as a request body, that indeed sounds like a bug in the test.

File: pkg/agent/protocol/http/proxy_test.go [hunk 3]
```diff
@@ -371,35 +313,54 @@ func Test_ProxyHandler(t *testing.T) {
...
+				t.Fatalf("building request to proxy: %v", err)
+			}
+
+			resp, err := http.DefaultClient.Do(req)
+			if err != nil {
+				t.Fatalf("making request to proxy: %v", err)
+			}
 
 			if tc.expectedStatus != resp.StatusCode {
 				t.Errorf("expected status code '%d' but '%d' received ", tc.expectedStatus, resp.StatusCode)
```
Comment (pablochacin) [inline pkg/agent/protocol/http/proxy_test.go:355]: As you are using `t.Fatal` above, shouldn't we use it uniformly at least in this test?

File: pkg/agent/protocol/http/proxy_test.go (hunk 1 above)
Comment (roobre) [reply pkg/agent/protocol/http/proxy_test.go:344]: If I understand the code correctly, the request we're performing here would be the client request, meaning the first request of the flow. Am I right? If this is the case, I believe the body for that request belongs to the test case (if we want to add it).

Comment (roobre) [reply pkg/agent/protocol/http/proxy_test.go]: I've renamed the fields to clarify these are used to control the upstream server response, and fixed the bug where they were being sent by the client https://github.com/grafana/xk6-disruptor/pull/308/commits/371bb82319354c8bcb97b0ef34cdc27d7cfded92 

Does this align with what you said?

**Review by pablochacin — APPROVED** [review]
LGTM

**Pull request merged**